File: src/main/java/aurocosh/divinefavor/common/registry/mappers/EntryMapper.java
package aurocosh.divinefavor.common.registry.mappers;

import aurocosh.divinefavor.common.lib.interfaces.IIndexedEntry;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.registries.IForgeRegistryEntry;

import java.util.Collections;
import java.util.List;
import java.util.Map;

public class EntryMapper<T extends IForgeRegistryEntry & IIndexedEntry> {
    private final BiMap<Integer, T> ids;
    private final BiMap<ResourceLocation, T> names;

    public EntryMapper() {
        ids = HashBiMap.create();
        names = HashBiMap.create();
    }

    public Map<Integer, T> getIdMap() {
        return Collections.unmodifiableMap(ids);
    }

    public Map<T, Integer> getIdMapinversed() {
        return Collections.unmodifiableMap(ids.inverse());
    }

    public Map<ResourceLocation, T> getNameMap() {
        return Collections.unmodifiableMap(names);
    }

    public Map<T, ResourceLocation> getNameMapinversed() {
        return Collections.unmodifiableMap(names.inverse());
    }

    public void register(List<T> values) {
        for (T value : values)
            register(value);
    }

    public void register(T value) {
        ids.put(value.getId(), value);
        names.put(value.getRegistryName(), value);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/registry/mappers/EntryMapper.kt
package aurocosh.divinefavor.common.registry.mappers

import aurocosh.divinefavor.common.lib.interfaces.IIndexedEntry
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import net.minecraft.util.ResourceLocation
import net.minecraftforge.registries.IForgeRegistryEntry
import java.util.*

class EntryMapper<T> where T : IForgeRegistryEntry<*>, T : IIndexedEntry {
    private val ids: BiMap<Int, T>
    private val names: BiMap<ResourceLocation, T>

    val idMap: Map<Int, T>
        get() = Collections.unmodifiableMap(ids)

    val idMapinversed: Map<T, Int>
        get() = Collections.unmodifiableMap(ids.inverse())

    val nameMap: Map<ResourceLocation, T>
        get() = Collections.unmodifiableMap(names)

    val nameMapinversed: Map<T, ResourceLocation>
        get() = Collections.unmodifiableMap(names.inverse())

    init {
        ids = HashBiMap.create()
        names = HashBiMap.create()
    }

    fun size(): Int {
        return ids.size
    }

    fun register(values: List<T>) {
        for (value in values)
            register(value)
    }

    fun register(value: T) {
        ids[value.id] = value
        names[value.registryName] = value
    }
}
